Extra picky

pohly opened the pull request:
This is needed for A/B partitioning. If you can, please review soon, because then I don't need to add unmerged patches to meta-swupd.

pohly commented:
@matthewrsj I tried to investigate the "make check" error, but I'm probably missing something in my local setup, because for me all tests fail. Are there instructions for running these tests?

icarus-sparry commented:
@pohly The tests are written using the [BATS](https://github.com/sstephenson/bats) testing framework. which is a set of bash scripts. You can run for example `test/functional/bundlelist/none-has-deps/test.bats` to just run a single test. Simplified, the test.bats file is run through 'sed' to process lines starting '@test' to convert these into bash functions, and then these functions are called in turn whilst bash has the '-e' flag set so anything which returns a non-zero return code causes the test to fail. `run` is a predefined function in this environment which captures the output and return code into variables and returns 0, so you can use that for things you expect to fail or you want to inspect.

pohly commented:
My tests failed because the swupd executable did not find libbsdiff.so.1 because it wasn't installed in the standard search path. Running the tests did not show that.

Regarding the test/functional/verify/fix-version-mismatch failure in Travis: that's because I changed the output that the test checks for in "lines-checked". Fixed.

Running test/functional/verify/fix-version-mismatch/test.bats printed a diff that isn't visible in the Travis build log. Perhaps bats should be invoked differently so that it dumps errors instead of storing them locally?

matthewrsj commented:
> Perhaps bats should be invoked differently so that it dumps errors instead of storing them locally?

This should have been visible in the travis build log. If any test fails the command `cat test-suite.log` is run, which should show that diff. I can't see the failed run in travis to identify if there was an issue with that or not.

phmccarty commented:
You can view the test-suite.log in Travis for failed builds by expanding the last section of the Travis build log. It would be nice to expand that section by default, but last time I looked into Travis configuration, I couldn't figure out how to get it working.

pohly commented:
On Mon, 2017-11-27 at 18:44 +0000, Patrick McCarty wrote:
> You can view the test-suite.log in Travis for failed builds by
> expanding the last section of the Travis build log. It would be nice
> to expand that section by default, but last time I looked into Travis
> configuration, I couldn't figure out how to get it working.

Indeed, the failure is there when one knows where to look:
https://travis-ci.org/clearlinux/swupd-client/builds/306780184#L5895

icarus-sparry commented on src/extra_files.c:
```diff
@@ -63,15 +56,34 @@ static inline int bsearch_helper(const void *A, const void *B)
+	char *savedname = strdup(relname); /* Only store name relative to top of area */
 	F = realloc(F, (nF + 1) * sizeof(*F));		      /* TODO, check realloc is smart, so don't need to double myself */
-	if (!F || !name) {
+	if (!F || !savedname) {
```
Good catch. thank you.

pohly commented on src/extra_files.c:
```diff
@@ -107,16 +119,12 @@ static void handle(const char *filename, bool is_dir, bool fix)
 
 /* expect the start to end in /usr and be the absolute path to the root */
-int walk_tree(struct manifest *manifest, const char *start, bool fix)
+int walk_tree(struct manifest *manifest, const char *start, bool fix, const regex_t *whitelist)
```
So you are saying that "whitelist" should be a global variable?

I personally don't consider that good style and tried to avoid it, hence the additional parameter to walk_tree(). Then for the callback there's no alternative, but at least it is static to extra_files.c and not global to all of swupd-client.

But I can change it.

pohly replied:
Thinking about this some more I still didn't like adding even more global variables to the entire program. This gets executes so infrequently (once per program run!) that the overhead is completely negligible.

Let me know whether this really needs to be changed.

icarus-sparry replied:
No, it doesn't need to be changed.

pohly commented on src/verify.c:
```diff
@@ -93,7 +100,9 @@ static void print_help(const char *name)
-	fprintf(stderr, "   -Y, --picky             List files which should not exist\n");
+	fprintf(stderr, "   -Y, --picky             List (without --fix) or remove (with --fix) files which should not exist (limited to /usr)\n");
+	fprintf(stderr, "   -X, --extra-picky       List (without --fix) or remove (with --fix) files which should not exist anywhere under /\n");
+	fprintf(stderr, "   -w, --picky-whitelist=re POSIX extended regular expression. Any file path matching the expression is ignored by --[extra-]picky. Matched directories get skipped completely. Empty regex matches nothing. Example: ^/var/|^/etc/machine-id$. Default: %s\n", picky_whitelist_default);
```
So the behavior is okay, it's just the description that needs to be changed? Can you suggest how to rephrase it?

matthewrsj replied:
I would suggest updating --picky and --extra-picky help strings to show what the default is, then say that -w overrides the default.

pohly commented on src/verify.c:
```diff
@@ -111,7 +120,16 @@ static void print_help(const char *name)
-
+	int errcode;
+
+	errcode = regcomp(&cmdline_option_picky_whitelist, picky_whitelist_default, REG_NOSUB|REG_EXTENDED);
```
Okay.

pohly replied:
Done.

pohly commented on src/verify.c:
```diff
@@ -111,7 +120,16 @@ static void print_help(const char *name)
+	errcode = regcomp(&cmdline_option_picky_whitelist, picky_whitelist_default, REG_NOSUB|REG_EXTENDED);
+	cmdline_option_picky_whitelist_set = true;
+	if (errcode) {
+		char buffer[160];
```
Okay.

pohly replied:
Done, with simple code (no loop, no if check) because performance is irrelevant here.

pohly commented on src/verify.c:
```diff
@@ -111,7 +120,16 @@ static void print_help(const char *name)
+		fprintf(stderr, "Invalid default --picky-whitelist: %s\n\n", buffer);
+		goto err;
+	}
 	while ((opt = getopt_long(argc, argv, "hxnItNbm:p:u:P:c:v:fYiF:qS:C:", prog_opts, NULL)) != -1) {
```
Okay.

pohly replied:
Done.

pohly commented on src/verify.c:
```diff
@@ -210,6 +228,31 @@ static bool parse_options(int argc, char **argv)
 		case 'Y':
 			cmdline_option_picky = true;
 			break;
+		case 'X':
```
I intentionally left out updating the man page, because --picky is also not documented and I didn't want to conflict with whatever someone else might be doing right now to fix that issue #335.

But I can add documentation for --picky and --extra-picky here, once we agree on what the actual options should be.

Do you want me to remove --picky entirely or add --picky-tree instead of the current --extra-picky?

icarus-sparry replied:
I was suggesting something like using '--picky' to enable the 'scan everything rather than just the files listed in the manifests', and '--picky-tree=/usr' and '--picky-tree=/' as options to be the 'restrict the effects of --picky to just parts of the tree'. Then what is currently '--picky' would be '--picky', and the new '--extra-picky' would be '--picky --picky-tree=/'.

matthewrsj replied:
I like @icarus-sparry's idea if it still works with @pohly's use-case.

pohly replied:
Yes, sounds good.

pohly replied:
I've implemented --picky-tree as proposed and updated the docs.

pohly commented on src/verify.c:
```diff
@@ -867,21 +911,23 @@ int verify_main(int argc, char **argv)
+			char *start = mk_full_filename(path_prefix, cmdline_option_extra_picky ? "" : "/usr");
+			fprintf(stderr, "%s removing extra files under %s\n", cmdline_option_extra_picky ? "--extra-picky" : "--picky", start);
+			ret = walk_tree(official_manifest, start, true,
+					cmdline_option_picky_whitelist_set ? &cmdline_option_picky_whitelist : NULL);
```
regex_t is a struct and there is no documented way of determining whether it was initialized, at least as far as I can tell. So that has to be tracked separately to determine whether the a regex was chosen.

I used the same naming scheme as for the other existing variables. Should I use something shorter even if that then becomes inconsistent with the rest of the code? I don't have any strong opinion either way.

icarus-sparry replied:
I disagree, cmdline_option_picky_whitelist is nothing like the other existing variables. It is the result of compiling a regular expression which might have been given on the command line, or might be the default that is hard coded into the program. All the others are reflect if an option had been given on the command line. You could use cmdline_option_picky_whitelist as the name of the variable that holds the optarg value in the 'w' case, i.e. the command line textual representation of the regular expression, and something much shorter, e.g. whitelist_re for a pointer to the compiled version of that.

I would set up storage for the regex_t, and then where you are setting cmdline_option_picky_whitelist_set to true/false I would set whitespace_re to &storage or NULL.

I think a lot of the problem I have is I find the whole handling of `-w` un-intuitive. If you don't give it then you preserve some things like ${XXX}/usr/lib/{modules,kernel}, if you do give it then it is the list of patterns of things to preserve, unless it is empty, in which case it says don't preserve anything.  This special case of the empty pattern just looks weird to me. I know from decades of experience that empty regular expressions match everything, not nothing. If I wanted to match nothing I would give it a pattern like ^$ or /does/not/match/anything. This contradicts the help message, which should say something like "if the pattern is empty then it disables the whitelist feature which defaults to %s, otherwise it replaces the is a regular expression of things to whitelist"

pohly replied:
The variable is now called "picky_whitelist" and it's a pointer that gets checked for NULL.

pohly commented:
Force-pushed the revised commit.

icarus-sparry commented on src/extra_files.c:
```diff
@@ -63,15 +56,34 @@ static inline int bsearch_helper(const void *A, const void *B)
+		return 0;
+	}
+
+	if (path_whitelist) {
```
path_whitelist will now always be set, so we can remove this conditional.

pohly replied:
It's not explicitly documented, but walk_tree() could be called with a NULL regex. It's not used as a general purpose utility function now, but it could be one day, so I'd prefer to not make assumptions about where the regex comes from and whether it's always set.

icarus-sparry commented on src/extra_files.c:
```diff
@@ -63,15 +56,34 @@ static inline int bsearch_helper(const void *A, const void *B)
+	if (path_whitelist) {
+		int match_res;
+		if (type == FTW_D) {
+			/* include trailing slash for regex like ^/var/ */
```
I still don't understand the motivation for this code.

You are adding a / to the end of directory names because you are asking the user to do the same, but what problem are you trying to solve?

icarus-sparry commented on src/verify.c:
```diff
@@ -108,11 +117,44 @@ static void print_help(const char *name)
 	int opt;
-
-	while ((opt = getopt_long(argc, argv, "hxnItNbm:p:u:P:c:v:fYiF:qS:C:", prog_opts, NULL)) != -1) {
+	if (!compile_whitelist(picky_whitelist_default)) {
```
You don't need to compile pick_whitelist_default. Change the name to say picky_whitelist_src,

in the 'w' case, assign optarg to picky_whitelist_src

move the compile of the regular expression until after the argument parsing has been completed.

pohly replied:
Okay.

pohly commented on docs/swupd.1.rst:
```diff
@@ -266,6 +266,39 @@ SUBCOMMANDS
+        ignored by --picky. Matched directories get skipped
+        completely. Directory paths end in a slash. This makes it
+        possible to write expressions that match a directory but not
+        something else that happens to have the same base name.
```
Yes, because directories are special I wanted to allow matching them separately. I also wanted to ensure that a regex like ^/usr/local does not accidentally also matches /usr/local.old (the mistake made in the current code).

But I agree, the former is a rather weak use case and the later is better solved by anchoring with $.

I'll revise the patch so that anchoring is enforced and without the trailing slash special case.

pohly commented on docs/swupd.1.rst:
```diff
@@ -266,6 +266,39 @@ SUBCOMMANDS
+
+            Ignores the ``/var`` directory and everything inside it
+            (because directories get skipped) and the
+            ``/etc/machine-id`` file.
```
In this original example, the /var symlink would not be matched. If that had been desired, ``^/var$`` would have to be part of the regex.

I'll change it so that ``/var = ^(/var)$`` matches both the /var dir and /var symlink.

pohly commented on src/verify.c:
```diff
@@ -108,11 +117,44 @@ static void print_help(const char *name)
+	if (picky_whitelist) {
+		regfree(picky_whitelist);
+	} else {
+		picky_whitelist = calloc(1, sizeof(*picky_whitelist));
```
Then we are back to "statically allocated struct + extra variable", which is what I had before for the same reasons that you mention here. Okay, I'll go back to that, with a pointer instead of bool for the extra variable.

pohly commented on src/verify.c:
```diff
@@ -108,11 +117,44 @@ static void print_help(const char *name)
+		}
+		regerror(errcode, picky_whitelist, buffer, len);
+		fprintf(stderr, "Invalid --picky-whitelist=%s: %s\n", regex, buffer);
+		return false;
```
Will fix.

pohly replied:
Regarding ``picky_whitelist = NULL``: the documentation does not explicitly say whether regfree() can or must be called after a failed regcomp(). My interpretation was that it needs to be called because regcomp() might have stored dynamically allocated error information in the regex_t for regerror(). Therefore not unsetting picky_whitelist is intentionally.

pohly commented on src/verify.c:
```diff
@@ -210,6 +252,22 @@ static bool parse_options(int argc, char **argv)
+			if (!optarg) {
+				fprintf(stderr, "Missing --picky-tree argument\n\n");
+				goto err;
+			}
```
Okay.

pohly replied:
I've used a slightly different error message with /usr as example, to ensure that users do not get confused whether the parameter must already include the --path prefix.

matthewrsj replied:
+1 for the error message.

pohly commented:
Revised patch pushed.

matthewrsj commented on src/verify.c:
```diff
@@ -108,11 +121,45 @@ static void print_help(const char *name)
+	/* Enforce matching the entire path. */
+	string_or_die(&full_regex, "^(%s)$", cmdline_option_picky_whitelist);
+
+	assert(!picky_whitelist);
```
Why fail here if picky_whitelist is set? You only call compile_whitelist at one point so I don't see how it could be set.

matthewrsj commented:
Please run `make compliant` on your branch to identify code style issues. These fixes are very minor but I would have to do it before the next release anyways.

This is a good change. LGTM except for the one question I had.

pohly commented:
On Wed, 2017-11-29 at 19:11 +0000, Matthew Johnson wrote:
> @matthewrsj commented on this pull request.
> +	assert(!picky_whitelist);
> Why fail here if picky_whitelist is set?

Because then memory would be leaked. Earlier I had code there that did
a regfree(), but as that code never would be executed now, I removed it
and replaced it with the assert().

>  You only call compile_whitelist at one point so I don't see how it
> could be set.

That's exactly the point of an assert: verify that an assumption is
correct.

As you said, the function only gets called once and thus
picky_whitelist should be NULL. The assert() documents that and (at
least in debug builds) causes a noticable failure when that assumption
is violated.

pohly commented:
On Wed, 2017-11-29 at 19:17 +0000, Matthew Johnson wrote:
> Please run make compliant on your branch to identify code style
> issues.

Done, force pushed. I ended up changing also one otherwise unmodified
line (the one with realloc) because that's what clang-format suggested.

> These fixes are very minor but I would have to do it before the next
> release anyways.

I've added another commit with a fix for something that I hadn't
touched.

With that commit, "make compliant" passes.

matthewrsj commented:
> The assert() documents that and (at least in debug builds) causes a noticable failure when that assumption is violated.

Alright.

> I've added another commit with a fix for something that I hadn't touched.

I've noticed that `make compliant` will sometimes catch new things when new code is added or old code is changed. Thanks for fixing the new error.

matthewrsj commented on src/verify.c:
```diff
@@ -210,6 +257,24 @@ static bool parse_options(int argc, char **argv)
+				goto err;
+			}
+			if (optarg[0] != '/') {
+				fprintf(stderr, "--picky-tree must be an absolute path, for example /usr");
```
Need a `\n` here. I'm seeing this:
```
--picky-tree must be an absolute path, for example /usrUsage:
   swupd verify [OPTION...]
```

pohly commented:
On Thu, 2017-11-30 at 18:03 +0000, Matthew Johnson wrote:
> +			if (optarg[0] != '/') {
> +				fprintf(stderr, "--picky-tree must
> be an absolute path, for example /usr");
> 
> Need a \n here. I'm seeing this:
> 
> --picky-tree must be an absolute path, for example /usrUsage:
>    swupd verify [OPTION...]

I've made that \n\n (same as in the other error messages). Force-
pushed.
> 
>   
>   
> 
> 
> 
> 
> 
> {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c5
> 5493e4bb","name":"GitHub"},"entity":{"external_key":"github/clearlinu
> x/swupd-client","title":"clearlinux/swupd-client","subtitle":"GitHub
> repository","main_image_url":"https://cloud.githubusercontent.com/ass
> ets/143418/17495839/a5054eac-5d88-11e6-95fc-
> 7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent
> .com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-
> b52498112777.png","action":{"name":"Open in
> GitHub","url":"https://github.com/clearlinux/swupd-
> client"}},"updates":{"snippets":[{"icon":"PERSON","message":"@matthew
> rsj requested changes on #337"}],"action":{"name":"View Pull
> Request","url":"https://github.com/clearlinux/swupd-
> client/pull/337#pullrequestreview-80275134"}}}

matthewrsj commented:
+1 from me. @icarus-sparry looks like you still have an outstanding review when I look at the `Reviewers` section. Do you have any further comments?

icarus-sparry commented:
I am happy

icarus-sparry commented:
I am happy.
+1
